### PR TITLE
[#5200]Update the Column data type for python client

### PR DESCRIPTION
**Pull request by SophieTech88**
### What changes were proposed in this pull request?
1. Implement type.java as type.py in Python client
2.  Implement types.java as types.py in Python client
3.  Create a test_types.py in unit tests.

### Why are the changes needed?
We need to support the column data type in python client

Fix: #5200

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Need to add the unit tests

### CODE REVIEW

**Comment by jerryshao** [comment]
@mchades can you please help to review?

**Comment by jerryshao** [comment]
@noidname01 , would you please also help to review this PR? Thanks a lot.

**Comment by SophieTech88** [comment]
When I double check the types folder (gravitino/tree/main/api/src/main/java/org/apache/gravitino/rel/types/) I find it has another file Decimal.java, so just update a new commit to convert decimal.java to decimal.python with unit test in python client folder.

File: clients/client-python/gravitino/api/decimal.py
Comment (mchades) [inline clients/client-python/gravitino/api/decimal.py:1]: missing license header

File: clients/client-python/gravitino/api/decimal.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+from decimal import Decimal as PyDecimal, ROUND_HALF_UP
+from typing import Union
+from .types import Types
+
+
+class Decimal:
```
Comment (mchades) [inline clients/client-python/gravitino/api/decimal.py:6]: Here is the reason why we need to wrap Java `BigDecimal`:
https://github.com/apache/gravitino/blob/6d05ec4b8bf278c2dd7a36af35589f07b220fcab/api/src/main/java/org/apache/gravitino/rel/types/Decimal.java#L27-L34

Does the `PyDecimal` have the same issue?

File: clients/client-python/tests/unittests/test_decimal.py
Comment (mchades) [inline clients/client-python/tests/unittests/test_decimal.py:1]: miss license header

File: clients/client-python/gravitino/api/decimal.py
Comment (SophieTech88) [reply clients/client-python/gravitino/api/decimal.py:1]: Many thanks. Just added the license header.

File: clients/client-python/gravitino/api/decimal.py (hunk 1 above)
Comment (SophieTech88) [reply clients/client-python/gravitino/api/decimal.py:23]: Thank you so much to point out the reason why we need to wrap Java BigDecimal. Then I just double checked for Python PyDecimal, it tends to handle this more intuitively for currency by counting all visible digits. So for example:
```
from decimal import Decimal

value = Decimal("0.01") ## it is 2
print("Precision:", value.adjusted() + 1)  ## it is 2 
print("Scale:", -value.as_tuple().exponent)
```
So PyDecimal didn't have the similar issue like Java BigDecimal. 
I will remove the new change for decimal.py.

File: clients/client-python/gravitino/api/types.py [hunk 2]
```diff
@@ -0,0 +1,921 @@
...
+class Types:
+
+    class NullType(Type):
+        _instance = None
+
+        @classmethod
+        def get(cls):
+            if cls._instance is None:
+                cls._instance = cls()
+            return cls._instance
```
Comment (mchades) [inline clients/client-python/gravitino/api/types.py:39]: I'm not sure if these codes can guarantee the singleton of `_instance` under race conditions in Python, can you help confirm it? @xunliu @noidname01

File: clients/client-python/gravitino/api/decimal.py (hunk 1 above)
Comment (SophieTech88) [reply clients/client-python/gravitino/api/decimal.py]: Removed. Done.

File: clients/client-python/gravitino/api/types.py (hunk 2 above)
Comment (SophieTech88) [reply clients/client-python/gravitino/api/types.py:39]: Thank you for catching that! You’re absolutely correct—this implementation does not currently handle race conditions, and under high concurrency, there’s a chance multiple threads could attempt to create the singleton instance simultaneously, breaking the singleton pattern.

Solution: Adding Thread Safety with a Lock
To handle this, we can use threading.Lock, which ensures that only one thread can enter the critical section where the instance is created. 

Here is the example:
```
class template:
    _instance = None
    _lock = threading.Lock()  # A lock to ensure thread safety

    @classmethod
    def get(cls):
        with cls._lock:  # Ensures only one thread can access this block at a time
            if cls._instance is None:
                cls._instance = cls()
        return cls._instance
 ```

Comment (mchades) [reply clients/client-python/gravitino/api/types.py:39]: `Lock` makes things more complex, is there another simple common method?

Comment (SophieTech88) [reply clients/client-python/gravitino/api/types.py:39]: Another option is use __new__():
```
class Types:

    class NullType(Type):
        _instance = None

        @classmethod
        def get(cls):
            if cls._instance is None:
                cls._instance = cls.__new__(cls)
            return cls._instance

```
logic here:
- get() checks if _instance is None.
- If _instance is None, cls.__new__(cls) is called to create a new instance of NullType without directly calling the constructor __init__.
- get() then returns the instance. Subsequent calls to get() will return the existing _instance.

Comment (jerryshao) [reply clients/client-python/gravitino/api/types.py:39]: Does Python support singleton?

**Review by xunliu — COMMENTED** [review]
hi @SophieTech88 
I help you improve this PR, please help me review my change parts.
1. We can use `__new__()` implement singleton.
2. Added some return type and function param type.
3. Added some class and function comments.

File: clients/client-python/gravitino/api/types.py [hunk 3]
```diff
@@ -0,0 +1,1106 @@
...
+        _instance: "BooleanType" = None
+
+        def __new__(cls):
+            if cls._instance is None:
+                cls._instance = super(Types.BooleanType, cls).__new__(cls)
+            return cls._instance
+
+        @classmethod
+        def get(cls) -> "BooleanType":
+            return cls()
```
Comment (mchades) [inline clients/client-python/gravitino/api/types.py:67]: should return `_instance` directly?

Comment (xunliu) [reply clients/client-python/gravitino/api/types.py:67]: `return cls()` will called `__new__()` function.

File: clients/client-python/gravitino/api/types.py (hunk 2 above)
Comment (SophieTech88) [reply clients/client-python/gravitino/api/types.py]: Yes, Python does support the Singleton pattern, which ensures that only one instance of a class is created. While Python does not enforce singletons natively, you can implement it in several ways.

File: clients/client-python/gravitino/api/types.py [hunk 4]
```diff
@@ -0,0 +1,1106 @@
...
+
+    class BooleanType(PrimitiveType):
+        """The boolean type in Gravitino."""
+
+        _instance: "BooleanType" = None
+
+        def __new__(cls):
+            if cls._instance is None:
+                cls._instance = super(Types.BooleanType, cls).__new__(cls)
+            return cls._instance
```
Comment (mchades) [inline clients/client-python/gravitino/api/types.py:63]: It appears that the code is not thread-safe.

**Comment by SophieTech88** [comment]
> hi @SophieTech88 I help you improve this PR, please help me review my change parts.
> 
> 1. We can use `__new__()` implement singleton.
> 2. Added some return type and function param type.
> 3. Added some class and function comments.

@xunliu Thank you so much. It is a great help.

File: clients/client-python/gravitino/api/types.py (hunk 4 above)
Comment (xunliu) [reply clients/client-python/gravitino/api/types.py:63]: @SophieTech88 Please add thread lock in the xxxType just like this, and need to add Test case `thread-safe`, thanks!

```
class BooleanType(PrimitiveType):
    """The boolean type in Gravitino."""

    _instance: "BooleanType" = None
    _lock = threading.Lock()

    def __new__(cls):
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super(BooleanType, cls).__new__(cls)
        return cls._instance
```

File: clients/client-python/tests/unittests/test_gvfs_with_local.py
Comment (mchades) [inline clients/client-python/tests/unittests/test_gvfs_with_local.py:1]: what are the codes for?

Comment (xunliu) [reply clients/client-python/tests/unittests/test_gvfs_with_local.py:1]: @mchades The `clients/client-python/tests/unittests/test_gvfs_with_local.py` report  waring `C0302: Too many lines in module (1102/1100) (too-many-lines)`, So I fixed it.

File: clients/client-python/gravitino/api/types.py (hunk 4 above)
Comment (xunliu) [reply clients/client-python/gravitino/api/types.py:63]: Yes, I agreed @diqiu50

**Review by mchades — APPROVED** [review]
The type definition logic LGTM

**Review by xunliu — APPROVED** [review]
@SophieTech88 Thank you for your contributions.
LGTM

**Review by mchades — APPROVED** [review]

**Pull request merged**